aws/src/main/java/org/apache/iceberg/aws/LakeFormationAwsClientFactory.java

```diff
@@ -0,0 +1,199 @@
+import software.amazon.awssdk.services.glue.model.GetTableResponse;
+import software.amazon.awssdk.services.kms.KmsClient;
+import software.amazon.awssdk.services.lakeformation.LakeFormationClient;
+import software.amazon.awssdk.services.lakeformation.model.GetTemporaryGlueTableCredentialsRequest;
+import software.amazon.awssdk.services.lakeformation.model.GetTemporaryGlueTableCredentialsResponse;
+import software.amazon.awssdk.services.lakeformation.model.Permission;
+import software.amazon.awssdk.services.s3.S3Client;
+import software.amazon.awssdk.services.sts.StsClient;
+import software.amazon.awssdk.services.sts.auth.StsAssumeRoleCredentialsProvider;
+import software.amazon.awssdk.services.sts.model.AssumeRoleRequest;
+import software.amazon.awssdk.services.sts.model.Tag;
+
+public class LakeFormationAwsClientFactory
+    implements AwsClientFactory {
+  private Map<String, String> properties;
+  private Region region;
+  private String s3Endpoint;
+
+  public LakeFormationAwsClientFactory() {
+  }
+
+  @Override
+  public void initialize(Map<String, String> catalogProperties) {
+    this.properties = Maps.newHashMap(catalogProperties);
+    this.region = region(properties);
+    this.s3Endpoint = properties.get(AwsProperties.S3FILEIO_ENDPOINT);
+  }
+
+  @Override
+  public S3Client s3() {
+    return S3Client.builder()
+        .endpointOverride(URI.create(s3Endpoint))
+        .credentialsProvider(buildS3CredentialProvider())
+        .region(region)
+        .httpClientBuilder(UrlConnectionHttpClient.builder())
+        .build();
+  }
+
+  @Override
+  public GlueClient glue() {
+    return GlueClient.builder()
+        .httpClientBuilder(UrlConnectionHttpClient.builder())
+        .build();
+  }
+
+  @Override
+  public KmsClient kms() {
+    return KmsClient.builder()
+        .httpClientBuilder(UrlConnectionHttpClient.builder())
+        .build();
+  }
+
+  @Override
+  public DynamoDbClient dynamo() {
+    return DynamoDbClient.builder()
+        .httpClientBuilder(UrlConnectionHttpClient.builder())
+        .build();
+  }
+
+  private AwsCredentialsProvider buildS3CredentialProvider() {
+
+    GlueClient glue = glue();
+    String glueCatalogId = properties.get(AwsProperties.GLUE_CATALOG_ID);
+    String databaseName = properties.get(AwsProperties.LAKE_FORMATION_DB_NAME);
+    String tableName = properties.get(AwsProperties.LAKE_FORMATION_TABLE_NAME);
+
+    GetTableResponse response = glue.getTable(GetTableRequest.builder()
+        .catalogId(glueCatalogId)
+        .databaseName(databaseName)
+        .name(tableName)
+        .build());
+    boolean isRegisteredWithLakeFormation = response.table().isRegisteredWithLakeFormation() != null ?
+        response.table().isRegisteredWithLakeFormation() : Boolean.FALSE;
+
+    if (isRegisteredWithLakeFormation) {
+      return new LakeFormationCredentialsProvider(properties);
+    } else {
+      return LakeFormationCredentialsProvider.buildAssumeRoleCredentialsProvider(properties);
```

Comment:
else shouldn't we just use normal default credential chain like all the other clients?

Reply:
it seems like we should not just have buildS3CredentialProvider, we need this whole logic in the parent method, because if `isRegisteredWithLakeFormation`, we not only need LF credential provider but also the assume role region. Otherwise we do not need credential or region.